src/DistributedComputations/distributed_transpose.jl

```diff
@@ -0,0 +1,113 @@
+using Oceananigans.Grids: architecture
+using Oceananigans.Architectures: on_architecture
+using KernelAbstractions: @index, @kernel
+using MPI: VBuffer, Alltoallv!
+
+# Transpose directions are assumed to work only in the following configuration
+# z -> y -> x -> y -> z
+# where z stands for z-complete data, y for y-complete data, and x for x-complete data
+# The initial field is always assumed to be in the z-complete configuration
+
+# Fallbacks for slab decompositions
+transpose_z_to_y!(::SlabYFields) = nothing
+transpose_y_to_z!(::SlabYFields) = nothing
+transpose_x_to_y!(::SlabXFields) = nothing
+transpose_y_to_x!(::SlabXFields) = nothing
+
+# Since z -> y -> x -> y -> z we only nedd to define the `pack` and `unpack` kernels
+# for the x and z configurations once, y requires two definitions depending on which
+# configuration it's interacting with
+
+@kernel function _pack_buffer_z!(yzbuff, zfield, N)
+    i, j, k = @index(Global, NTuple)
+    @inbounds yzbuff.send[j + N[2] * (i-1 + N[1] * (k-1))] = zfield[i, j, k]
+end
+
+@kernel function _pack_buffer_x!(xybuff, xfield, N)
+    i, j, k = @index(Global, NTuple)
+    @inbounds xybuff.send[j + N[2] * (k-1 + N[3] * (i-1))] = xfield[i, j, k]
+end
+
+@kernel function _pack_buffer_yx!(xybuff, yfield, N) # y -> x
+    i, j, k = @index(Global, NTuple)
+    @inbounds xybuff.send[i + N[1] * (k-1 + N[3] * (j-1))] = yfield[i, j, k]
+end
+
+@kernel function _pack_buffer_yz!(xybuff, yfield, N) # y -> z
+    i, j, k = @index(Global, NTuple)
+    @inbounds xybuff.send[k + N[3] * (i-1 + N[1] * (j-1))] = yfield[i, j, k]
+end
+
+@kernel function _unpack_buffer_x!(xybuff, xfield, N, n)
+    i, j, k = @index(Global, NTuple)
+    nm = n[1], N[2], N[3]
+    @inbounds begin
+        i′  = mod(i - 1, nm[1]) + 1
+        m   = (i - 1) ÷ nm[1]
+        idx = i′ + nm[1] * (k-1 + nm[3] * (j-1)) + m * prod(nm)
+        xfield[i, j, k] = xybuff.recv[idx]
+    end
+end
+
+@kernel function _unpack_buffer_z!(yzbuff, zfield, N, n)
+    i, j, k = @index(Global, NTuple)
+    nm = N[1], N[2], n[3]
+    @inbounds begin
+        k′  = mod(k - 1, nm[3]) + 1
+        m   = (k - 1) ÷ nm[3]
+        idx = k′ + nm[3] * (i-1 + nm[1] * (j-1)) + m * prod(nm)
+        zfield[i, j, k] = yzbuff.recv[idx]
+    end
+end
+
+@kernel function _unpack_buffer_yz!(yzbuff, yfield, N, n) # z -> y
+    i, j, k = @index(Global, NTuple)
+    nm = N[1], n[2], N[3]
+    @inbounds begin
+        j′  = mod(j - 1, nm[2]) + 1
+        m   = (j - 1) ÷ nm[2]
+        idx = j′ + nm[2] * (i-1 + nm[1] * (k-1)) + m * prod(nm)
+        yfield[i, j, k] = yzbuff.recv[idx]
+    end
+end
+
+@kernel function _unpack_buffer_yx!(yzbuff, yfield, N, n) # x -> y
```

Comment:
in the notation above, "pack buffer yx" means "pack a buffer for transposing from y to x".

But in this notation it seems we are filling the `yfield`. So for "unpacking", the notation is inverted and "unpack buffer yx" means "unpack buffer from x to y"? Do I understand this right? I'm confused a little bit

Reply:
I can rethink the names. But in essence, `_pack_buffer_x!` and `_pack_buffer_z!` are packing the buffer from a `x-complete` configuration and from a `z-complete` configuration. There is no ambiguity here because the x- and z- configuration communicate only through the y-configuration. On the other hand, for the y-configuration there are two ways to pack a buffer and two ways to unpack it depending on whether the y-configuration is going (or coming from) a x- or a z-configuration

Reply:
> There is no ambiguity here because the x- and z- configuration communicate only through the y-configuration.

Maybe there is no ambiguity if you understand the entire code, but to understand the meaning of these names you have to also understand the entire algorithm (ie this function does not make sense in isolation if you only read this one function). It could be nice to have the code understandable if we just read chunks of it, rather than requiring us to understand the entire functionality in order to just understand these few lines...

Reply:
Just flagging this as an unresolved comment.

Reply:
I have made functions name very explicit. This should be resolved